design/transaction-layer-refactoring.md

```diff
@@ -0,0 +1,151 @@
+# Transaction Layer Refactoring
+
+## Motivation
+
+At the very beginning, all transactional commands in TiKV share common procedures:
+
+1. Acquire latches
+2. Check constraints and generate modifications
+3. Apply modifications to the raft store
+4. Reply to RPC and release the latches
+
+After the latches and the snapshot are acquired, all the commands depend on nothing else, so we needn't care about passing dependencies in deeply.
+
+So the current implementation structure is natural and great: different commands are wrapped as enum variants and go through the whole process. In each step we match the type of command and do specific work.
+
+However, as more commands and optimizations are added, this is not always the case. Some commands breaks the procedure and adds dependencies.
+
+For example, the "pipelined pessimistic lock" replies before the raft store finishes replication, which is not following the common procedure. Also, when an `AcquirePessimisticLock` command meets a lock, it needs to wait for the lock being released in TiKV for a while. And when a lock is released, the awaiting `AcquirePessimisticLock` commands are notified. These are distinctive steps that are not shared by other commands. More dependencies (the lock manager and the deadlock detector) are also introduced.
+
+If more and more commands are not sharing the same procedure and dependencies, the current structure will have no benefit. Instead, the drawbacks become clearer.
+
+The code for a certain command is spread in too many files. Each function handles a part of work of various commands. People are hard to understand what happens about a single command.
+
+And now, different commands have different dependencies. But because all commands share the same procedure, we must pass in the union of all dependencies of all commands. It makes adding commands and dependencies more difficult.
+
+## New structure
+
+Based on the analysis above, the current structure is not flexible enough and should be dropped. Instead, we can change to put logic of a single command together. Common steps can be extracted as methods for reuse. Then, it will be much easier to find out the procedure of each command while still not repeating code.
+
+Steps like acquiring latches and waiting for raft replication cannot finish immediately, so it is appropriate to make every command an `async fn`.
+
+For example, the `AcquirePessimisticLock` command can be like:
+
+```rust
+async fn acquire_pessimistic_lock(&self, req: PessimisticLockRequest) -> PessimisticLockResponse {
+    let mut resp = PessimisticLockResponse::default();
+    let guards = self.lock_keys(req.mutations.iter().map(|m| m.get_key())).await;

```

Comment:
Can there be some way to protect us from forgetting lock latches?

Reply:
It's just a demo. I think we can start it as soon as possible?

Reply:
@nrc @longfangsong what do you think?

Reply:
@MyonKeminta I can't come up with a good way to prevent but I think it's also hard to forget now.
People start to add new commands by copying the code of other commands. In this design, the code of a command has only one place, so it's hard to forget locking latches.